pyOCD/target/memory_map.py

```diff
@@ -0,0 +1,165 @@
+"""
+ mbed CMSIS-DAP debugger
+ Copyright (c) 2015 ARM Limited
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from xml.etree import ElementTree
+
+__all__ = ['MemoryRegion', 'MemoryMap', 'RamRegion', 'RomRegion', 'FlashRegion']
+
+MAP_XML_HEADER =  """<?xml version="1.0"?>
+<!DOCTYPE memory-map PUBLIC "+//IDN gnu.org//DTD GDB Memory Map V1.0//EN" "http://sourceware.org/gdb/gdb-memory-map.dtd">
+"""
+
+## @brief One contiguous range of memory.
+class MemoryRegion(object):
+    def __init__(self, type='ram', start=0, end=0, length=0, blocksize=0, name='', isBootMemory=False):
+        self._type = type
+        self._start = start
+        if length != 0:
+            self._end = self._start + length - 1
+        else:
+            self._end = end
+        self._blocksize = blocksize
+        if not name:
+            self._name = self._type
+        else:
+            self._name = name
+        self._is_boot_mem = isBootMemory
+
+    @property
+    def type(self):
+        return self._type
+
+    @property
+    def start(self):
+        return self._start
+
+    @property
+    def end(self):
+        return self._end
+
+    @property
+    def length(self):
+        return self._end - self._start + 1
+
+    @property
+    def blocksize(self):
+        return self._blocksize
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def isFlash(self):
+        return self._type == 'flash'
+
+    @property
+    def isRam(self):
+        return self._type == 'ram'
+
+    @property
+    def isRom(self):
+        return self._type == 'rom'
+
+    @property
+    def isBootMemory(self):
+        return self._is_boot_mem
+
```

Comment:
Another property that would be good to have is isPoweredOnBoot.  That way the debugger knows to avoid things like external ram/flash that hasn't been setup yet but is in the memory map.  It would also allow test scripts to verify all of the powered regions can be read from.